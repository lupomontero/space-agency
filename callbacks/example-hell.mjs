import { buildRocket, fetchCrew, getFuel, bookLaunchPad } from './index.mjs';

export const run = (cb) => {
  const start = Date.now();
  buildRocket((err, rocket) => {
    if (err) {
      return cb(err);
    }
    fetchCrew((err, crew) => {
      if (err) {
        return cb(err);
      }
      getFuel((err, fuel) => {
        if (err) {
          return cb(err);
        }
        bookLaunchPad((err, launchPad) => {
          if (err) {
            return cb(err);
          }

          launchPad.addRocket(rocket, (err) => {
            if (err) {
              return cb(err);
            }
            rocket.addCrew(crew, (err) => {
              if (err) {
                return cb(err);
              }
              rocket.addFuel(fuel, (err) => {
                if (err) {
                  return cb(err);
                }
                const fire = () => {
                  if (rocket.altitude < 100) {
                    rocket.fireEngine(0, (err) => {
                      if (err) {
                        return cb(err);
                      }
                      fire();
                    });
                  } else if (rocket.altitude < 400) {
                    if (rocket.engineIsFiring(0)) {
                      rocket.stopEngine(0, (err) => {
                        if (err) {
                          return cb(err);
                        }
                        rocket.runStageSeparation(0, (err) => {
                          if (err) {
                            return cb(err);
                          }
                          fire();
                        });
                      });
                    } else {
                      rocket.fireEngine(1, (err) => {
                        if (err) {
                          return cb(err);
                        }
                        fire();
                      });
                    }
                  } else {
                    rocket.stopEngine(1, (err) => {
                      if (err) {
                        return cb(err);
                      }
                      rocket.runStageSeparation(1, (err) => {
                        if (err) {
                          return cb(err);
                        }
                        console.log(`Total time: ${Math.floor((Date.now() - start) / 1000)} secs`);
                        cb();
                      });
                    });
                  }
                };
                fire();
              });
            });
          });
        });
      });
    });
  });
};


if (process.env.NODE_ENV !== 'test') {
  run((err) => {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
}
