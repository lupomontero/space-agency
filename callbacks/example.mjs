import { buildRocket, fetchCrew, getFuel, bookLaunchPad } from './index.mjs';

const concurrent = (obj, cb) => {
  let hasErrored = false;
  const results = {};
  const keys = Object.keys(obj);
  const handler = name => (err, data) => {
    if (hasErrored) {
      return;
    }
    if (err) {
      hasErrored = true;
      return cb(err);
    }
    Object.assign(results, { [name]: data });
    if (Object.keys(results).length === keys.length) {
      cb(null, results);
    }
  };
  keys.forEach((key) => {
    obj[key](handler(key));
  });
};

const series = (fns, cb) => {
  const recurse = (remaining, memo = []) => {
    if (!remaining.length) {
      return cb(null, memo);
    }
    remaining[0]((err, data) => {
      if (err) {
        return cb(err);
      }
      return recurse(remaining.slice(1), memo.concat(data));
    });
  };
  return recurse(fns);
};

const initMission = cb => concurrent({
  rocket: buildRocket,
  crew: fetchCrew,
  fuel: getFuel,
  launchPad: bookLaunchPad,
}, cb);

const getReadyForLaunch = ({ rocket, crew, fuel, launchPad }, cb) => {
  series([
    launchPad.addRocket.bind(null, rocket),
    rocket.addCrew.bind(null, crew),
    rocket.addFuel.bind(null, fuel),
  ], cb);
};

const launch = ({ rocket, crew, fuel, launchPad }, cb) => {
  const fire = (err) => {
    if (err) {
      return cb(err);
    }
    if (rocket.altitude < 100) {
      return rocket.fireEngine(0, fire);
    }
    if (rocket.altitude < 400 && rocket.engineIsFiring(0)) {
      return series([
        rocket.stopEngine.bind(null, 0),
        rocket.runStageSeparation.bind(null, 0),
      ], fire);
    }
    if (rocket.altitude < 400) {
      return rocket.fireEngine(1, fire);
    }
    return series([
      rocket.stopEngine.bind(null, 1),
      rocket.runStageSeparation.bind(null, 1),
    ], cb);
  };
  fire();
};

export const run = (cb) => {
  const start = Date.now();
  return initMission((err, mission) => {
    if (err) {
      return cb(err);
    }
    return series([
      getReadyForLaunch.bind(null, mission),
      launch.bind(null, mission),
    ], (err) => {
      if (err) {
        return cb(err);
      }
      console.log(`Total time: ${Math.floor((Date.now() - start) / 1000)} secs`);
      cb();
    });
  });
};


if (process.env.NODE_ENV !== 'test') {
  run((err) => {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
}
