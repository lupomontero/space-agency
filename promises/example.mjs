import { buildRocket, fetchCrew, getFuel, bookLaunchPad } from './index.mjs';

const initMission = () => Promise.all([
  buildRocket(),
  fetchCrew(),
  getFuel(),
  bookLaunchPad(),
]);

const getReadyForLaunch = ([rocket, crew, fuel, launchPad]) => (
  launchPad.addRocket(rocket)
    .then(() => rocket.addCrew(crew))
    .then(() => rocket.addFuel(fuel))
    .then(() => rocket)
);

const launch = rocket => new Promise((resolve, reject) => {
  const fire = () => {
    if (rocket.altitude < 100) {
      return rocket.fireEngine(0).then(fire, reject);
    }
    if (rocket.altitude < 400 && rocket.engineIsFiring(0)) {
      return rocket.stopEngine(0)
        .then(() => rocket.runStageSeparation(0))
        .then(fire, reject);
    }
    if (rocket.altitude < 400) {
      return rocket.fireEngine(1).then(fire, reject);
    }
    return rocket.stopEngine(1)
      .then(() => rocket.runStageSeparation(1))
      .then(resolve, reject);
  };
  fire();
});

export const run = () => {
  const start = Date.now();
  return initMission()
    .then(getReadyForLaunch)
    .then(launch)
    .then(() => {
      console.log(`Total time: ${Math.floor((Date.now() - start) / 1000)} secs`);
    });
};


if (process.env.NODE_ENV !== 'test') {
  run()
    .catch((err) => {
      if (err) {
        console.error(err);
        process.exit(1);
      }
    });
}
