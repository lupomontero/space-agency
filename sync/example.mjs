import { buildRocket, fetchCrew, getFuel, bookLaunchPad } from './index.mjs';

const initMission = () => {
  const rocket = buildRocket();
  const crew = fetchCrew();
  const fuel = getFuel();
  const launchPad = bookLaunchPad();
  return { rocket, crew, fuel, launchPad };
};

export const run = () => {
  const start = Date.now();
  const { rocket, crew, fuel, launchPad } = initMission();

  launchPad.addRocket(rocket);

  rocket.addCrew(crew);
  rocket.addFuel(fuel);

  while (rocket.altitude < 100) {
    rocket.fireEngine(0);
  }

  while (rocket.altitude < 400) {
    if (rocket.engineIsFiring(0)) {
      rocket.stopEngine(0);
      rocket.runStageSeparation(0);
    }
    rocket.fireEngine(1);
  }

  rocket.stopEngine(1);
  rocket.runStageSeparation(1);

  console.log(`Total time: ${Math.floor((Date.now() - start) / 1000)} secs`);
};


if (typeof process !== 'undefined' && process.env.NODE_ENV !== 'test') {
  run();
}
