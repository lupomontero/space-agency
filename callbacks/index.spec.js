import { buildRocket, fetchCrew, getFuel, bookLaunchPad } from './index.mjs';
import { run } from './example.mjs';
import { run as runHell } from './example-hell.mjs';

let consoleLogSpy;

beforeEach(() => {
  consoleLogSpy = jest.spyOn(global.console, 'log').mockImplementation(jest.fn());
});

afterEach(() => {
  consoleLogSpy.mockRestore();
});

describe('buildRocket', () => {
  it('should return a rocket in 5 seconds', (done) => {
    const start = Date.now();
    buildRocket((err, rocket) => {
      expect(err).toBeFalsy();
      expect(typeof rocket.addCrew).toBe('function');
      expect(typeof rocket.addFuel).toBe('function');
      expect(typeof rocket.fireEngine).toBe('function');
      expect(typeof rocket.stopEngine).toBe('function');
      expect(typeof rocket.engineIsFiring).toBe('function');
      expect(typeof rocket.runStageSeparation).toBe('function');
      expect(Math.floor((Date.now() - start) / 1000)).toBe(5);
      expect(consoleLogSpy.mock.calls).toEqual([
        ['Building rocket...'],
        ['Rocket built!'],
      ]);
      done();
    });
  });
});

describe('fetchCrew', () => {
  it('should return a crew array in 3 seconds', (done) => {
    const start = Date.now();
    fetchCrew((err, crew) => {
      expect(err).toBeFalsy();
      expect(Array.isArray(crew)).toBe(true);
      expect(crew.length).toBe(2);
      expect(Math.floor((Date.now() - start) / 1000)).toBe(3);
      done();
    });
  });
});

describe('getFuel', () => {
  it('should return fuel in 1 second', (done) => {
    const start = Date.now();
    getFuel((err, fuel) => {
      expect(err).toBeFalsy();
      expect(typeof fuel).toBe('object');
      expect(Math.floor((Date.now() - start) / 1000)).toBe(1);
      done();
    });
  });
});

describe('bookLaunchPad', () => {
  it('should return a launchPad object in 5 seconds', (done) => {
    const start = Date.now();
    bookLaunchPad((err, launchPad) => {
      expect(err).toBeFalsy();
      expect(typeof launchPad.addRocket).toBe('function');
      expect(Math.floor((Date.now() - start) / 1000)).toBe(5);
      done();
    });
  });
});

describe('integration', () => {
  it('should run the whole mission in 11 seconds!', (done) => {
    const start = Date.now();
    run((err) => {
      expect(err).toBeFalsy();
      expect(Math.floor((Date.now() - start) / 1000)).toBe(11);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot();
      done();
    });
  }, 30 * 1000);
});

describe('integration (hell)', () => {
  it('should run the whole mission in 11 seconds!', (done) => {
    const start = Date.now();
    run((err) => {
      expect(err).toBeFalsy();
      expect(Math.floor((Date.now() - start) / 1000)).toBe(11);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot();
      done();
    });
  }, 30 * 1000);
});

