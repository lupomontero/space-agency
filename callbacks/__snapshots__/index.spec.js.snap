// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration (hell) should run the whole mission in 20 seconds! 1`] = `
Array [
  Array [
    "initMission",
  ],
  Array [
    "Building rocket...",
  ],
  Array [
    "Rocket built!",
  ],
  Array [
    "Fetching crew...",
  ],
  Array [
    "Crew ready!",
  ],
  Array [
    "Getting fuel...",
  ],
  Array [
    "Fuel ready!",
  ],
  Array [
    "Booking launch pad...",
  ],
  Array [
    "Launch pad booked!",
  ],
  Array [
    Object {
      "addCrew": [Function],
      "addFuel": [Function],
      "altitude": 400,
      "engineIsFiring": [Function],
      "engines": Array [
        Object {
          "isFiring": false,
        },
        Object {
          "isFiring": false,
        },
      ],
      "fireEngine": [Function],
      "runStageSeparation": [Function],
      "speed": 28000,
      "stopEngine": [Function],
    },
    Array [
      Object {
        "name": "Pepito Pérez",
      },
      Object {
        "name": "Manolita Fernández",
      },
    ],
    Object {},
    Object {
      "addRocket": [Function],
    },
  ],
  Array [
    "Adding rocket to launch pad...",
  ],
  Array [
    "Added rocket to launch pad!",
  ],
  Array [
    "Adding crew...",
  ],
  Array [
    "Crew ready!",
  ],
  Array [
    "Adding fuel...",
  ],
  Array [
    "Fuel ready!",
  ],
  Array [
    "Firing engine 0...",
  ],
  Array [
    "Altitude 10km | Speed 700km/h",
  ],
  Array [
    "Altitude 20km | Speed 1400km/h",
  ],
  Array [
    "Altitude 30km | Speed 2100km/h",
  ],
  Array [
    "Altitude 40km | Speed 2800km/h",
  ],
  Array [
    "Altitude 50km | Speed 3500km/h",
  ],
  Array [
    "Altitude 60km | Speed 4200km/h",
  ],
  Array [
    "Altitude 70km | Speed 4900km/h",
  ],
  Array [
    "Altitude 80km | Speed 5600km/h",
  ],
  Array [
    "Altitude 90km | Speed 6300km/h",
  ],
  Array [
    "Altitude 100km | Speed 7000km/h",
  ],
  Array [
    "Engine 0 cut-off...",
  ],
  Array [
    "Engine 0 cut-off completed...",
  ],
  Array [
    "Running stage 0 separation...",
  ],
  Array [
    "Stage 0 separation completed...",
  ],
  Array [
    "Firing engine 1...",
  ],
  Array [
    "Altitude 110km | Speed 7700km/h",
  ],
  Array [
    "Altitude 120km | Speed 8400km/h",
  ],
  Array [
    "Altitude 130km | Speed 9100km/h",
  ],
  Array [
    "Altitude 140km | Speed 9800km/h",
  ],
  Array [
    "Altitude 150km | Speed 10500km/h",
  ],
  Array [
    "Altitude 160km | Speed 11200km/h",
  ],
  Array [
    "Altitude 170km | Speed 11900km/h",
  ],
  Array [
    "Altitude 180km | Speed 12600km/h",
  ],
  Array [
    "Altitude 190km | Speed 13300km/h",
  ],
  Array [
    "Altitude 200km | Speed 14000km/h",
  ],
  Array [
    "Altitude 210km | Speed 14700km/h",
  ],
  Array [
    "Altitude 220km | Speed 15400km/h",
  ],
  Array [
    "Altitude 230km | Speed 16100km/h",
  ],
  Array [
    "Altitude 240km | Speed 16800km/h",
  ],
  Array [
    "Altitude 250km | Speed 17500km/h",
  ],
  Array [
    "Altitude 260km | Speed 18200km/h",
  ],
  Array [
    "Altitude 270km | Speed 18900km/h",
  ],
  Array [
    "Altitude 280km | Speed 19600km/h",
  ],
  Array [
    "Altitude 290km | Speed 20300km/h",
  ],
  Array [
    "Altitude 300km | Speed 21000km/h",
  ],
  Array [
    "Altitude 310km | Speed 21700km/h",
  ],
  Array [
    "Altitude 320km | Speed 22400km/h",
  ],
  Array [
    "Altitude 330km | Speed 23100km/h",
  ],
  Array [
    "Altitude 340km | Speed 23800km/h",
  ],
  Array [
    "Altitude 350km | Speed 24500km/h",
  ],
  Array [
    "Altitude 360km | Speed 25200km/h",
  ],
  Array [
    "Altitude 370km | Speed 25900km/h",
  ],
  Array [
    "Altitude 380km | Speed 26600km/h",
  ],
  Array [
    "Altitude 390km | Speed 27300km/h",
  ],
  Array [
    "Altitude 400km | Speed 28000km/h",
  ],
  Array [
    "Engine 1 cut-off...",
  ],
  Array [
    "Engine 1 cut-off completed...",
  ],
  Array [
    "Running stage 1 separation...",
  ],
  Array [
    "Stage 1 separation completed...",
  ],
  Array [
    "Total time: 20 secs",
  ],
]
`;

exports[`integration should run the whole mission in 11 seconds! 1`] = `
Array [
  Array [
    "Building rocket...",
  ],
  Array [
    "Fetching crew...",
  ],
  Array [
    "Getting fuel...",
  ],
  Array [
    "Booking launch pad...",
  ],
  Array [
    "Fuel ready!",
  ],
  Array [
    "Crew ready!",
  ],
  Array [
    "Rocket built!",
  ],
  Array [
    "Launch pad booked!",
  ],
  Array [
    "Adding rocket to launch pad...",
  ],
  Array [
    "Added rocket to launch pad!",
  ],
  Array [
    "Adding crew...",
  ],
  Array [
    "Crew ready!",
  ],
  Array [
    "Adding fuel...",
  ],
  Array [
    "Fuel ready!",
  ],
  Array [
    "Firing engine 0...",
  ],
  Array [
    "Altitude 10km | Speed 700km/h",
  ],
  Array [
    "Altitude 20km | Speed 1400km/h",
  ],
  Array [
    "Altitude 30km | Speed 2100km/h",
  ],
  Array [
    "Altitude 40km | Speed 2800km/h",
  ],
  Array [
    "Altitude 50km | Speed 3500km/h",
  ],
  Array [
    "Altitude 60km | Speed 4200km/h",
  ],
  Array [
    "Altitude 70km | Speed 4900km/h",
  ],
  Array [
    "Altitude 80km | Speed 5600km/h",
  ],
  Array [
    "Altitude 90km | Speed 6300km/h",
  ],
  Array [
    "Altitude 100km | Speed 7000km/h",
  ],
  Array [
    "Engine 0 cut-off...",
  ],
  Array [
    "Engine 0 cut-off completed...",
  ],
  Array [
    "Running stage 0 separation...",
  ],
  Array [
    "Stage 0 separation completed...",
  ],
  Array [
    "Firing engine 1...",
  ],
  Array [
    "Altitude 110km | Speed 7700km/h",
  ],
  Array [
    "Altitude 120km | Speed 8400km/h",
  ],
  Array [
    "Altitude 130km | Speed 9100km/h",
  ],
  Array [
    "Altitude 140km | Speed 9800km/h",
  ],
  Array [
    "Altitude 150km | Speed 10500km/h",
  ],
  Array [
    "Altitude 160km | Speed 11200km/h",
  ],
  Array [
    "Altitude 170km | Speed 11900km/h",
  ],
  Array [
    "Altitude 180km | Speed 12600km/h",
  ],
  Array [
    "Altitude 190km | Speed 13300km/h",
  ],
  Array [
    "Altitude 200km | Speed 14000km/h",
  ],
  Array [
    "Altitude 210km | Speed 14700km/h",
  ],
  Array [
    "Altitude 220km | Speed 15400km/h",
  ],
  Array [
    "Altitude 230km | Speed 16100km/h",
  ],
  Array [
    "Altitude 240km | Speed 16800km/h",
  ],
  Array [
    "Altitude 250km | Speed 17500km/h",
  ],
  Array [
    "Altitude 260km | Speed 18200km/h",
  ],
  Array [
    "Altitude 270km | Speed 18900km/h",
  ],
  Array [
    "Altitude 280km | Speed 19600km/h",
  ],
  Array [
    "Altitude 290km | Speed 20300km/h",
  ],
  Array [
    "Altitude 300km | Speed 21000km/h",
  ],
  Array [
    "Altitude 310km | Speed 21700km/h",
  ],
  Array [
    "Altitude 320km | Speed 22400km/h",
  ],
  Array [
    "Altitude 330km | Speed 23100km/h",
  ],
  Array [
    "Altitude 340km | Speed 23800km/h",
  ],
  Array [
    "Altitude 350km | Speed 24500km/h",
  ],
  Array [
    "Altitude 360km | Speed 25200km/h",
  ],
  Array [
    "Altitude 370km | Speed 25900km/h",
  ],
  Array [
    "Altitude 380km | Speed 26600km/h",
  ],
  Array [
    "Altitude 390km | Speed 27300km/h",
  ],
  Array [
    "Altitude 400km | Speed 28000km/h",
  ],
  Array [
    "Engine 1 cut-off...",
  ],
  Array [
    "Engine 1 cut-off completed...",
  ],
  Array [
    "Running stage 1 separation...",
  ],
  Array [
    "Stage 1 separation completed...",
  ],
  Array [
    "Total time: 11 secs",
  ],
]
`;
